ioctl/test_ioctl.c

```diff
@@ -0,0 +1,348 @@
+	TEST_ASSERT_EQUAL_INT32(1, flag);
+}
+
+
+TEST(ioctl, in_val)
+{
+	int32_t rdata, data = 14;
+	int ret;
+
+	/* Send data to driver by value */
+	ret = ioctl(dd, TIOCVAL, data);
+	TEST_ASSERT_EQUAL_INT(0, ret);
+	TEST_ASSERT_EQUAL_INT(0, lseek(fd, 0, SEEK_SET));
+	TEST_ASSERT_EQUAL_INT(4, read(fd, &rdata, 4));
+	TEST_ASSERT_EQUAL_INT32(14, rdata);
+}
+
+
+TEST(ioctl, data_in)
+{
+	int32_t rdata, data_in = 20;
+	int ret;
+
+	/* Send data to driver by pointer */
+	ret = ioctl(dd, TIOCPTR, &data_in);
+	TEST_ASSERT_EQUAL_INT(0, ret);
+	TEST_ASSERT_EQUAL_INT(0, lseek(fd, 0, SEEK_SET));
+	TEST_ASSERT_EQUAL_INT(4, read(fd, &rdata, 4));
+	TEST_ASSERT_EQUAL_INT32(20, rdata);
+}
+
+
+TEST(ioctl, data_in_big)
+{
+	/* Big data so that it is not copied inside message */
+	int8_t data_in[BIG_SIZE];
+	memset(data_in, 3, BIG_SIZE);
+	int8_t rdata[BIG_SIZE];
+	int ret;
+
+	/* Send data to driver by pointer , data big enough to not be copied by ioctl_pack directly into message */
+	ret = ioctl(dd, TIOCPTR2, data_in);
+	TEST_ASSERT_EQUAL_INT(0, ret);
+	TEST_ASSERT_EQUAL_INT(0, lseek(fd, 0, SEEK_SET));
+	TEST_ASSERT_EQUAL_INT(BIG_SIZE, read(fd, rdata, BIG_SIZE));
+	TEST_ASSERT_EQUAL_MEMORY(data_in, rdata, BIG_SIZE);
+}
+
+
+TEST(ioctl, data_out)
+{
+	int32_t data_out;
+	int ret;
+
+	/* Get data from driver */
+	ret = ioctl(dd, TIOCOUT, &data_out);
+	TEST_ASSERT_EQUAL_INT(0, ret);
+	TEST_ASSERT_EQUAL_INT32(15, data_out);
+}
+
+
+TEST(ioctl, data_out_big)
+{
+	/* Big data so that it is not copied inside message */
+	int8_t data_expected[BIG_SIZE];
+	memset(data_expected, 5, BIG_SIZE);
+	int8_t data_out[BIG_SIZE];
+	int ret;
+
+	/* Get data from driver */
+	ret = ioctl(dd, TIOCOUT2, data_out);
+	TEST_ASSERT_EQUAL_INT(0, ret);
+	TEST_ASSERT_EQUAL_MEMORY(data_expected, data_out, BIG_SIZE);
+}
+
+
+TEST_GROUP_RUNNER(ioctl)
+{
+	char devName[5] = "test";
+	oid_t dev;
+	int res, n = 0;
+
+	fd = open(PATH, O_RDWR | O_CREAT | O_TRUNC);
+	if (fd < 0) {
+		fprintf(stderr, "Couldn't open file\n");
+		return 1;
+	}
+
+	res = portCreate(&port);
+	if (res != EOK) {
+		fprintf(stderr, "Couldn't create port\n");
+		return 1;
+	}
+
+	dev.port = port;
+	dev.id = 0;
+
+	res = create_dev(&dev, devName);
+	if (res != EOK) {
+		fprintf(stderr, "Coldn't create device\n");
+		return 1;
+	}
+
+	res = pthread_create(&tid, NULL, thread, NULL);
+	if (res != 0) {
+		fprintf(stderr, "Couldn't create thread\n");
+		return 1;
+	}
+
+	while ((dd = open("/dev/test", O_RDWR)) < 0 && n < MAX_FAIL) {
+		usleep(10000);
+		n++;
+	}
+
+	if (n == MAX_FAIL) {
+		close(fd);
+		remove(PATH);
+		fprintf(stderr, "Can't open device file\n");
+		return 1;
+	}
+	/* main thread - run test cases */
+	//RUN_TEST_CASE(ioctl, regular_file);
+	RUN_TEST_CASE(ioctl, wrong_fd);
+	RUN_TEST_CASE(ioctl, no_data);
+	RUN_TEST_CASE(ioctl, in_val);
+	RUN_TEST_CASE(ioctl, data_in);
+	RUN_TEST_CASE(ioctl, data_in_big);
+	RUN_TEST_CASE(ioctl, data_out);
+	RUN_TEST_CASE(ioctl, data_out_big);
```

Comment:
You can also add a case with passing some integers as arg (now it's not checked):
![Screenshot from 2023-05-31 11-43-17](https://github.com/phoenix-rtos/phoenix-rtos-tests/assets/77304431/61b7a86b-5086-4f75-a005-c8e241d41ffd)

Reply:
You can certainly add a case, where you send request, which is not handled.